fabric-api-lookup-api-v1/src/main/java/net/fabricmc/fabric/api/lookup/v1/entity/EntityApiLookup.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.lookup.v1.entity;
+
+import java.util.function.BiFunction;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.entity.Entity;
+import net.minecraft.entity.EntityType;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.impl.lookup.entity.EntityApiLookupImpl;
+
+/**
+ * An object that allows retrieving APIs from entities.
+ * Instances of this interface can be obtained through {@link #get}
+ *
+ * <p>When trying to {@link #find} an API for an entity, the provider registered for the entity type will be queried if it exists.
+ * If it doesn't exist, or if it returns {@code null}, the fallback providers will be queried in order.
+ *
+ * <p><h3>Usage Example</h3>
+ * Let's pretend that we have the following interface that we want to attach to entities.
+ * <pre>{@code
+ * public interface Leveled {
+ *     int getLevel();
+ * }
+ * }</pre>
+ *
+ * <p>We need to create the EntityApiLookup. We don't need any context so we use {@link Void}.
+ * <pre>{@code
+ * public class MyApi {
+ *     public static final EntityApiLookup<Leveled, Void> LEVELED_ENTITY = EntityApiLookup.get(new Identifier("mymod:leveled_entity"), Leveled.class, Void.class);
+ * }
+ * }</pre>
+ *
+ * <p>Now we can query instances of {@code Leveled}.
+ * <pre>{@code
+ * Leveled leveled = MyApi.LEVELED_ENTITY.find(entity, null);
+ * if (leveled != null) {
+ *     // Do something with the API.
+ *     System.out.println("Entity " + entity.getEntityName() + " is level " + leveled.getLevel());
+ * }
+ * }</pre>
+ *
+ * <p>For query to return useful result, we must expose the API.
+ * <pre>{@code
+ * // If the entity directly implements the interface, registerSelf can be used.
+ * public class LeveledPigEntity extends PigEntity implements Leveled {
+ *     ...
+ * }
+ * MyApi.LEVELED_ENTITY.registerSelf(LEVELED_PIG_ENTITY_TYPE);
+ *
+ * // Otherwise, registerForTypes can be used.
+ * MyApi.LEVELED_ENTITY.registerForTypes((entity, ignored) -> {
+ *     // Return a Leveled instance for your entity here, or null if there's none.
+ *     // The context is Void in this case, so it can be ignored.
+ * }, EntityType.PLAYER, EntityType.ZOMBIE);
```

Comment:
Should probably use `registerForType` as the example as it will allow using the entity subclass directly (I know this isn't done in BlockApiLookup, if you want to fix it there as well why not).